odoo/http.py

```diff
@@ -1012,18 +1011,29 @@ def authenticate(self, db, login=None, password=None):
             REMOTE_ADDR=wsgienv['REMOTE_ADDR'],
         )
         uid = odoo.registry(db)['res.users'].authenticate(db, login, password, env)
+        user = request.env(user=uid)['res.users'].browse(uid)
+        self.pre_uid = uid
+        self.pre_session_token = user._compute_session_token(self.sid)
+
+        if not user._mfa_url():
+            self.finalize()
 
         self.rotate = True
         self.db = db
-        self.uid = uid
         self.login = login
-        self.session_token = security.compute_session_token(self, request.env)
-        request.uid = uid
         request.disable_db = False
 
-        self.get_context()
         return uid
 
+    def finalize(self):
+        """ Finalizes a partial session, should be called on MFA validation to
+        convert a partial / pre-session into a full-fledged "logged-in" one
+        """
+        self.rotate = True
+        request.uid = self.uid = self.pop('pre_uid')
+        self.session_token = self.pop('pre_session_token')
```

Comment:
TODO: should we regenerate the session token (and check that it matches the pre-session) instead of just copying over? /cc @tbe-odoo

Reply:
Hello, I think it is safer to regen the session token, someone might override the `_get_session_token_fields` in `res.users` and add a field that potentially changes during the full authentication.

Reply:
Should I also check that the old and new session tokens still match or does that not matter?

Reply:
I don't think we need to check the old one, they shouldn't be different. So your actual solution is working  (if none of the session token fields change) but I think it is preferable to generate a new one in case of a change in-between.

Here is a possible case.
A piece of code decide to change a `session token field` and update the `session_token` value in the session to avoid a session invalidation.
If we use the `pre_session_token` which isn't updated, the session would be broken and the user might be unable to authenticate.

So checking the old session token might not be a good solution because this would invalidate the session which is not the intended behavior by the developer.
So I would generate a new session token and let the next request decide if the session is still valid

Reply:
in that case the pre_session_token doesn't seem very useful? I was storing it because it's the one generated during the first phase of the login and so that's the one for which we'd verified the login / password, but if we're not keeping it I can just remove the entire thing.

Reply:
Yup I think we can use the `session_token` directly